File: 11_mca_v2.py
"""
11 — MCA v2: respecified geometry of the supplier space
=======================================================
Project: Who Sells to the State (2026_2). GAN V2 remedy.

- Actives = endowments only: personería, rubro modal, modal client type.
- Supplementary = trajectory categories (intensity, tenure, directa) and
  provinces, projected as barycenters of their suppliers.
- Benzécri-corrected inertia; silhouette scan for k.
- Fracture test: separate MCAs per province stratum (metro / intermediate /
  peripheral); congruence of dim-1 category structure across strata.
- Provincial centroid trajectories per era: dispersion of centroids as the
  geometric counterpart of the Theil between-share.

Input:  data/processed/adjudicaciones_tipo.parquet
        data/processed/supplier_master.parquet
Output: tables/tab_mca2_benzecri.csv, tab_mca2_congruencia.csv,
        tab_mca2_centroides_provincia_era.csv
        figures/fig_mca2_espacio.png

Usage:  python 11_mca_v2.py
"""
from pathlib import Path

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import prince
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score

PROJECT = Path(__file__).parent
PROC = PROJECT / "data" / "processed"
TAB = PROJECT / "tables"
FIG = PROJECT / "figures"

NEA = {"Misiones", "Corrientes", "Chaco", "Formosa"}
METRO = {"CABA", "Buenos Aires", "Córdoba", "Santa Fe", "Mendoza"}
PERIF = NEA | {"Catamarca", "Jujuy", "La Rioja", "Salta",
               "Santiago del Estero", "Tucumán"}
ERAS = ["macri", "fernandez", "milei"]

adj = pd.read_parquet(PROC / "adjudicaciones_tipo.parquet")
adj = adj[adj["es_nuevo"] & adj["provincia"].notna()]
master = pd.read_parquet(PROC / "supplier_master.parquet")

cliente_modal = (adj.groupby("cuit")["tipo_organismo"]
                 .agg(lambda s: s.mode().iat[0]).rename("cliente_modal"))
m = (master[master["adjudicado"] & master["provincia"].notna()]
     .merge(cliente_modal, on="cuit", how="left").dropna(subset=["cliente_modal"]))

m["estrato"] = np.select([m["provincia"].isin(METRO), m["provincia"].isin(PERIF)],
                         ["metro", "periferica"], default="intermedia")


def band_personeria(p):
    return {"Persona Fisica": "PersFisica", "Sociedad Anonima": "SA",
            "Sociedad Responsabilidad Limitada": "SRL"}.get(p, "OtraForma")


top_rubros = m["rubro_modal"].value_counts().head(8).index
m["A_personeria"] = m["personeria"].map(band_personeria)
m["A_rubro"] = np.where(m["rubro_modal"].isin(top_rubros),
                        m["rubro_modal"].str.slice(0, 18), "OTROS").astype(str)
m["A_cliente"] = m["cliente_modal"]
ACTIVE = ["A_personeria", "A_rubro", "A_cliente"]

m["S_intensidad"] = pd.cut(m["n_adjudicaciones"], [0, 1, 5, 20, np.inf],
                           labels=["1adj", "2-5adj", "6-20adj", "20+adj"]).astype(str)
m["S_tenure"] = pd.cut(m["ultimo_ejercicio"] - m["primer_ejercicio"], [-1, 0, 3, np.inf],
                       labels=["1anio", "2-4anios", "5+anios"]).astype(str)
m["S_directa"] = pd.cut(m["share_directa"], [-0.01, 0.001, 0.5, 1.0],
                        labels=["sin_directa", "mixto", "mayoria_directa"]).astype(str)
SUPP = ["S_intensidad", "S_tenure", "S_directa"]


def benzecri(raw_eig, K):
    thr = 1.0 / K
    adj_e = [((K / (K - 1)) * (e - thr)) ** 2 for e in raw_eig if e > thr]
    total = sum(adj_e)
    return [e / total for e in adj_e]


def fit_mca(df):
    mca = prince.MCA(n_components=5, random_state=42).fit(df[ACTIVE])
    rows = mca.row_coordinates(df[ACTIVE])
    rows.columns = [f"dim{i+1}" for i in range(rows.shape[1])]
    return mca, rows


mca, coords = fit_mca(m)
raw_eig = list(mca.eigenvalues_)
bz = benzecri(raw_eig, K=len(ACTIVE))
pd.DataFrame({"dim": range(1, len(raw_eig) + 1), "eigenvalue": raw_eig,
              "pct_raw": [e / sum(raw_eig) for e in raw_eig],
              "pct_benzecri": bz + [np.nan] * (len(raw_eig) - len(bz))}
             ).to_csv(TAB / "tab_mca2_benzecri.csv", index=False)
print(f"Benzécri: dim1 {bz[0]:.1%}, dim2 {bz[1]:.1%}" if len(bz) > 1
      else f"Benzécri: dim1 {bz[0]:.1%}")

sil = {}
for k in range(3, 9):
    lab = KMeans(n_clusters=k, n_init=10, random_state=42).fit_predict(coords[["dim1", "dim2"]])
    sil[k] = silhouette_score(coords[["dim1", "dim2"]].sample(min(5000, len(coords)),
                                                              random_state=1),
                              pd.Series(lab, index=coords.index)
                              .loc[coords.sample(min(5000, len(coords)),
                                                 random_state=1).index])
best_k = max(sil, key=sil.get)
print("silhouette:", {k: round(v, 3) for k, v in sil.items()}, "| best k =", best_k)
m["cluster"] = KMeans(n_clusters=best_k, n_init=10, random_state=42) \
    .fit_predict(coords[["dim1", "dim2"]])

# ── Supplementary projections: barycenters ───────────────────────────────────
def barycenters(var):
    return (coords[["dim1", "dim2"]].groupby(m[var].values).mean()
            .assign(variable=var))


supp_pts = pd.concat([barycenters(v) for v in SUPP + ["provincia"]])

# provincial centroids per era (suppliers active in each era)
act = (adj.groupby(["cuit", "era"]).size().rename("n").reset_index())
act = act.merge(coords.assign(cuit=m["cuit"].values), on="cuit", how="inner") \
         .merge(m[["cuit", "provincia"]], on="cuit", how="left")
cent = (act.groupby(["provincia", "era"])[["dim1", "dim2"]].mean().reset_index())
disp = (cent.groupby("era")[["dim1", "dim2"]]
        .apply(lambda g: np.sqrt(g.var(ddof=0).sum())).rename("dispersion_centroides"))
cent.to_csv(TAB / "tab_mca2_centroides_provincia_era.csv", index=False)

# ── Fracture test: congruence of dim1 across strata ──────────────────────────
def cat_coords_dim1(df):
    mca_s = prince.MCA(n_components=3, random_state=42).fit(df[ACTIVE])
    cc = mca_s.column_coordinates(df[ACTIVE])[0]
    ev = benzecri(list(mca_s.eigenvalues_), K=len(ACTIVE))
    return cc, ev[0] if ev else np.nan


base_cc, _ = cat_coords_dim1(m[m["estrato"] == "metro"])
rows = []
for estrato in ["metro", "intermedia", "periferica"]:
    cc, bz1 = cat_coords_dim1(m[m["estrato"] == estrato])
    common = base_cc.index.intersection(cc.index)
    r = np.corrcoef(base_cc[common], cc[common])[0, 1]
    congr = abs(r)  # axis sign is arbitrary
    rows.append({"estrato": estrato, "n": (m["estrato"] == estrato).sum(),
                 "benzecri_dim1": round(bz1, 3),
                 "congruencia_dim1_vs_metro": round(congr, 3)})
cg = pd.DataFrame(rows)
cg.to_csv(TAB / "tab_mca2_congruencia.csv", index=False)

# ── Figure ───────────────────────────────────────────────────────────────────
fig, ax = plt.subplots(figsize=(10, 8))
ax.scatter(coords["dim1"], coords["dim2"], s=4, alpha=0.15, color="#bbbbbb")
cols = mca.column_coordinates(m[ACTIVE])
for cat, r in cols.iterrows():
    ax.annotate(str(cat)[:26], (r[0], r[1]), fontsize=7, color="#1f4e79", weight="bold")
for prov, r in supp_pts[supp_pts["variable"] == "provincia"].iterrows():
    is_nea = prov in NEA
    ax.scatter(r["dim1"], r["dim2"], marker="s",
               s=70 if is_nea else 28, color="#d62728" if is_nea else "#e8a0a0",
               zorder=5)
    if is_nea or prov in {"CABA", "Buenos Aires", "Córdoba"}:
        ax.annotate(prov, (r["dim1"], r["dim2"]), fontsize=8,
                    color="#a00", xytext=(4, 4), textcoords="offset points")
for era, marker in zip(ERAS, ["o", "D", "*"]):
    d = cent[cent["provincia"].isin(NEA) & (cent["era"] == era)]
    ax.scatter(d["dim1"], d["dim2"], marker=marker, s=60, facecolors="none",
               edgecolors="#7b241c", label=f"centroides NEA {era}")
ax.axhline(0, color="grey", lw=0.5)
ax.axvline(0, color="grey", lw=0.5)
ax.set_xlabel(f"Dim 1 ({bz[0]:.0%} Benzécri)")
ax.set_ylabel(f"Dim 2 ({bz[1]:.0%} Benzécri)" if len(bz) > 1 else "Dim 2")
ax.set_title("El espacio social de los proveedores del Estado — categorías activas,\n"
             "provincias como puntos suplementarios (NEA rojo)")
ax.legend(fontsize=7)
fig.tight_layout()
fig.savefig(FIG / "fig_mca2_espacio.png", dpi=200)
plt.close(fig)

print("\nDispersión de centroides provinciales por era (contraparte geométrica del Theil):")
print(disp.reindex(ERAS).round(4).to_string())
print("\nCongruencia dim1 vs metro (fracture test):")
print(cg.to_string(index=False))
print("\nCentroides NEA por era (dim1):")
print(cent[cent["provincia"].isin(NEA)].pivot_table(index="provincia",
      columns="era", values="dim1").reindex(sorted(NEA))[ERAS].round(3).to_string())
print("OK -> tables/, figures/")

File: 11_mca_v2.R
# 11 — The geometry of the supplier space (canonical R version).
# Upgrades over the Python pass: GDAtools::speMCA (Le Roux & Rouanet
# orthodoxy), supplementary variables with test values, and REAL
# class-specific MCA (csMCA: subcloud within the reference space) for the
# fracture test — the Python per-stratum MCAs were an approximation.
# Run: Rscript 11_mca_v2.R

source("theme_house.R", chdir = TRUE)
suppressPackageStartupMessages({library(GDAtools); library(cluster)})

set.seed(42)

adj <- read_parquet(file.path(DIR_PROC, "adjudicaciones_tipo.parquet"))
master <- read_parquet(file.path(DIR_PROC, "supplier_master.parquet"))
adj_win <- adj |> filter(es_nuevo, !is.na(provincia),
                         between(ejercicio, WIN0, WIN1))

m <- build_supplier_space(adj, master)
X <- active_matrix(m)
cat(sprintf("Espacio %d-%d: %d proveedores, %d categorías activas\n",
            WIN0, WIN1, nrow(m), sum(sapply(X, nlevels))))

RARE <- rare_categories(X)
cat("Categorías pasivadas (<5%):", paste(RARE, collapse = ", "), "\n")
mca <- speMCA(X, excl = excl_index(X, RARE), ncp = 5)
mr <- modif.rate(mca)
mrate <- mr$modif$mrate
benz <- tibble(dim = seq_len(nrow(mr$raw)),
               eigenvalue = mr$raw$eigen,
               pct_raw = mr$raw$rate,
               pct_benzecri = c(mrate, rep(NA, nrow(mr$raw) - length(mrate))))
write_tab(benz, "tab_mca2_benzecri.csv")
cat(sprintf("Benzécri: dim1 %.1f%%, dim2 %.1f%%\n", mrate[1], mrate[2]))

coords <- as.data.frame(mca$ind$coord)[, 1:2] |>
  setNames(c("dim1", "dim2")) |> as_tibble()
m <- bind_cols(m, coords)

# active category coordinates and contributions: the table the Results section
# reads the axes from (signs of MCA axes are arbitrary and can flip between
# runs, so the poles must be re-read from here after every re-run)
# category counts come from the data, not from the model: in a specific MCA
# `var$weight` is a relative weight over the retained categories and cannot be
# read as a frequency
conteos <- unlist(lapply(names(X), function(v) {
  setNames(as.integer(table(X[[v]])), paste0(v, ".", levels(X[[v]])))
}))
cat_tab <- as.data.frame(mca$var$coord)[, 1:2] |>
  setNames(c("dim1", "dim2")) |>
  tibble::rownames_to_column("categoria") |>
  mutate(variable = sub("\\..*$", "", categoria),
         ctr_dim1 = round(mca$var$contrib[, 1], 2),
         ctr_dim2 = round(mca$var$contrib[, 2], 2),
         n = unname(conteos[categoria]),
         pct = round(100 * n / nrow(X), 1)) |>
  arrange(dim1)
stopifnot(!any(is.na(cat_tab$n)))
write_tab(cat_tab |> mutate(across(where(is.numeric), ~round(.x, 3))),
          "tab_mca2_categorias.csv")

# silhouette scan
idx <- sample(nrow(m), min(5000, nrow(m)))
sil <- sapply(3:8, function(k) {
  km <- kmeans(coords, centers = k, nstart = 10)
  mean(silhouette(km$cluster[idx], dist(coords[idx, ]))[, 3])
})
best_k <- (3:8)[which.max(sil)]
cat("silhouette:", paste(3:8, round(sil, 3), sep = "=", collapse = " "),
    "| best k =", best_k, "\n")
m$cluster <- kmeans(coords, centers = best_k, nstart = 10)$cluster

# ── Supplementary variables: coords + test values ────────────────────────────
sup_tab <- bind_rows(lapply(c("S_intensidad", "S_tenure",
                              "S_cohorte", "S_fundamento",
                              "provincia", "estrato"),
                            function(v) {
  sv <- suppressWarnings(varsup(mca, factor(m[[v]])))
  as.data.frame(sv$coord)[, 1:2] |>
    setNames(c("dim1", "dim2")) |>
    mutate(variable = v, categoria = rownames(sv$coord),
           typic_dim1 = as.data.frame(sv$typic)[, 1],
           typic_dim2 = as.data.frame(sv$typic)[, 2])
}))
write_tab(sup_tab, "tab_mca2_supvars.csv")

# ── Provincial centroids per era + dispersion ────────────────────────────────
act <- adj_win |> distinct(cuit, era) |>
  inner_join(m |> select(cuit, provincia, dim1, dim2), by = "cuit")
cent <- act |> group_by(provincia, era) |>
  summarise(dim1 = mean(dim1), dim2 = mean(dim2), .groups = "drop")
write_tab(cent, "tab_mca2_centroides_provincia_era.csv")
disp <- cent |> group_by(era) |>
  summarise(dispersion_centroides =
              sqrt(sum(c(var(dim1), var(dim2)) * (n() - 1) / n())),
            .groups = "drop") |>
  mutate(era = factor(era, levels = ERAS)) |> arrange(era)

# ── Fracture test: REAL csMCA per stratum, congruence of dim1 ────────────────
cs_coords <- function(sel) {
  res <- csMCA(X, subcloud = sel, excl = excl_index(X, RARE), ncp = 3)
  cc <- as.data.frame(res$var$coord)[, 1:3]
  rownames(cc) <- rownames(res$var$coord)
  list(coord = cc, mrate = modif.rate(res)$modif$mrate[1])
}
base <- cs_coords(m$estrato == "core")
# axes of a specific MCA can reorder across subclouds: congruence per metro
# axis = the best absolute correlation over the stratum's dims 1-3, with the
# matching dim reported (full matrix kept in the table).
cg <- bind_rows(lapply(c("core", "middle", "outer"), function(s) {
  r <- cs_coords(m$estrato == s)
  common <- intersect(rownames(base$coord), rownames(r$coord))
  cmat <- abs(cor(base$coord[common, ], r$coord[common, ]))
  tibble(estrato = s, n = sum(m$estrato == s),
         benzecri_dim1 = round(r$mrate, 3),
         congruencia_dim1_bestmatch = round(max(cmat[1, ]), 3),
         dim_que_matchea = which.max(cmat[1, ]),
         congruencia_dim1_vs_dim1 = round(cmat[1, 1], 3),
         congruencia_dim2_bestmatch = round(max(cmat[2, ]), 3))
}))
write_tab(cg, "tab_mca2_congruencia.csv")

# ── Figure ───────────────────────────────────────────────────────────────────
cats <- as.data.frame(mca$var$coord)[, 1:2] |>
  setNames(c("dim1", "dim2")) |>
  mutate(categoria = rownames(mca$var$coord))
provs <- sup_tab |> filter(variable == "provincia")
cent_nea <- cent |> filter(provincia %in% NEA) |>
  mutate(era = factor(era, levels = ERAS))
m$region <- factor(region_of(m$provincia), levels = REGION_LEVELS)

p <- ggplot(m |> slice_sample(n = 8000), aes(dim1, dim2)) +
  geom_hline(yintercept = 0, colour = "grey70", linewidth = 0.3) +
  geom_vline(xintercept = 0, colour = "grey70", linewidth = 0.3) +
  geom_point(size = 0.4, alpha = 0.12, colour = "grey60") +
  stat_ellipse(aes(colour = region), level = 0.5, linewidth = 0.6) +
  geom_text(data = cats, aes(label = substr(categoria, 1, 24)),
            size = 2.5, colour = "#1f4e79", fontface = "bold") +
  geom_point(data = provs, aes(shape = categoria %in% NEA),
             size = 2, colour = "#d62728") +
  ggrepel::geom_text_repel(
    data = provs |> filter(categoria %in% c(NEA, "CABA", "Buenos Aires",
                                            "Córdoba")),
    aes(label = categoria), size = 2.4, colour = "#a00000", seed = 42) +
  geom_point(data = cent_nea, aes(fill = era), shape = 21, size = 2.4,
             colour = "#7b241c") +
  scale_colour_manual(values = REGION_COLS) +
  scale_shape_manual(values = c(`TRUE` = 15, `FALSE` = 0), guide = "none") +
  labs(x = sprintf("Dim 1 (%.0f%% Benzécri)", mr$mrate[1]),
       y = sprintf("Dim 2 (%.0f%% Benzécri)", mr$mrate[2]),
       title = "El espacio social de los proveedores del Estado (speMCA)",
       subtitle = "Categorías activas; provincias suplementarias; elipses de concentración por región; centroides NEA por era") +
  theme_house()
# diagnostic render, not a submission figure: it goes to the archive so that
# figures/ holds only the twelve figures of the manuscript
save_fig(p, "../_archive/figures_working/fig_mca2_espacio.png",
         width = 10, height = 8)

cat("\nDispersión de centroides por era:\n"); print(disp)
cat("\nCongruencia dim1 vs metro (csMCA real):\n"); print(cg)
cat("\nSuplementarias con |typic| máximo en dim2 (lectura del eje):\n")
print(sup_tab |> filter(variable != "provincia") |>
        arrange(desc(abs(typic_dim2))) |> head(6) |>
        mutate(across(where(is.numeric), ~round(.x, 2))))
